### ISSUE TITLE
proof of concept for offline PDF access

### ISSUE THREAD

**Issue by eroux** [buda-base/public-digital-library#734]
This is quite an important access mechanism that we'll have to develop. In the main use case, a user has a hard drive connected to their computer and the idea is that when the user accesses some images on BUDA, instead of opening the image viewer, the PDF in the local hard drive is opened.

Here's a proposal that we can test:
- we could create a small .html file that would go at the root of the directory where the PDFs are
- the html file would open an iframe of BUDA, with a special argument
- the BUDA iframe would send some postMessage to the html file instead of opening the image viewer
- the html file would then open the PDF corresponding to the messages it receives

@berger-n can you make a small proof of concept of that? (if it can work, fingers crossed!)

**Comment by berger-n** [comment]
- [x] not sure we need the `postMessage` thing if we simply redirect to BUDA with path of html file as url parameter: http://library-dev.bdrc.io/scripts/offlinePDFaccess.html (only redirects for now)
- [ ] then we can save it to `redux` state and/or `localStorage` as well in case user reloads page (I guess it will require a "disconnect from local drive" button somewhere)
- [ ] then whenever viewer is opened we simply don't use mirador but an iframe with local url to pdf instead, keeping current navigation menu to go back to BUDA (or actually rebuilding a simplified version of it, as it is part of mirador rendering I think)

**Comment by berger-n** [comment]
really in the proof-of-concept state for now, but working:

![Kazam_screencast_00023 mp4](https://user-images.githubusercontent.com/33383857/182403106-0b3fde5d-ec09-4729-b0a5-d72aa9d8a517.gif)

we'll also need a volume selector in that case, like a popup in BUDA?

**Comment by eroux** [comment]
Excellent, thanks a lot!

**Comment by eroux** [comment]
Yes we indeed need a volume selector, perhaps in the top html page? Or on BUDA if it's easier

**Comment by berger-n** [comment]
working online here (only for W12827, only ~80mb for the 5 volumes):  [link](http://library-dev.bdrc.io/scripts/offlinePDFaccess.html?host=library-dev.bdrc.io&url=http%3A%2F%2Flibrary-dev.bdrc.io%2Fsearch%3Fq%3D%2522snying%2520thig%2520ya%2520bzhi%2522~1%26lg%3Dbo-x-ewts%26t%3DInstance%26f%3Dasset%2Cinc%2Ctmp%3ApossibleAccess%26f%3Dasset%2Cinc%2Ctmp%3AcatalogOnly)
you can open pdf from results, version page, scans page -- only volume 1 for now
you can also reload and still be on the same page (though there are obviously some issues to fix with back/next buttons of browser)

![Kazam_screencast_00024 mp4](https://user-images.githubusercontent.com/33383857/182663602-49baa9b6-1f47-4096-8e8f-5fee43c0a05e.gif)
--

what about links to images in outline? I don't think we can open pdf at a given page (but let me check)

![image](https://user-images.githubusercontent.com/33383857/182663894-aed92f25-bf05-4755-9b54-24d908327e5a.png)

**Comment by eroux** [comment]
amazing, excellent job!! I think some browsers have that feature yes (opening a PDF at a certain page), at least Firefox does

**Comment by eroux** [comment]
@berger-n can you prioritize this? I'll create a full json file and will attache it here tomorrow

**Comment by eroux** [comment]
Here's the json

[DLD_2018.json.zip](https://github.com/buda-base/public-digital-library/files/9314716/DLD_2018.json.zip)

the format is more simple than I anticipated:

```json
{
  "W00EGS1016181": "09-ta-ལོ་རྒྱུས་དང་རྣམ་ཐར་སོགས།/11-གནས་ཡིག/W00EGS1016181-ཤར་ཕྱོགས་དུང་རི་དང་དགོན་སྡེ་དུ་མའི་དཀར་ཆག་མཐོང་བ་དོན་ལྡན།",
}
```

means that volume 1 will be in

```
09-ta-ལོ་རྒྱུས་དང་རྣམ་ཐར་སོགས།/11-གནས་ཡིག/W00EGS1016181-ཤར་ཕྱོགས་དུང་རི་དང་དགོན་སྡེ་དུ་མའི་དཀར་ཆག་མཐོང་བ་དོན་ལྡན།/W00EGS1016181-001.pdf
```

etc. all the volumes can be constructed according to that pattern.

**Comment by eroux** [comment]
so, now two issues:
- what do we do with records that are not on the DLD but are found on BUDA? I think we want to have a clear distinction when the user accesses the images. In a first iteration this could be a popup: "This record is not on your local DLD drive, the images will be consulted on the BUDA website directly". Or perhaps through an icon?
- what to do with files that are on the DLD but are not supposed to be accessible on BUDA? Do we pretend they're not on the DLD? @JannTibetan

**Comment by berger-n** [comment]
now with a volume selector (only 5 first volumes are already uploaded on dev server though): http://library-dev.bdrc.io/scripts/offlinePDFaccess.html?host=library-dev.bdrc.io&url=http%3A%2F%2Flibrary-dev.bdrc.io%2Fshow%2Fbdr%3AMW12827%3Fs%3Dq%253D%252522snying%252520thig%252520ya%252520bzhi%252522~1%2526lg%253Dbo-x-ewts%2526t%253DInstance%2526f%253Dasset%252Cinc%252Ctmp%253ApossibleAccess%2526f%253Dasset%252Cinc%252Ctmp%253AcatalogOnly%2526useDLD%253Dtrue%2526n%253D1

![image](https://user-images.githubusercontent.com/33383857/184359370-4ee4e72c-6113-4426-86f8-9e740a1e2f05.png)
---

here's an archive to be extracted at the root of the DLD
with an index.html that should work out of the box

[00-BUDA from DLD.zip](https://github.com/buda-base/public-digital-library/files/9316449/00-BUDA.from.DLD.zip)

**Comment by eroux** [comment]
Here are the final lists, there should be a selector for users to pick the DLD they're on:

[dld-lists.zip](https://github.com/buda-base/public-digital-library/files/9349737/dld-lists.zip)


A few things to be mindful of:
- W2KG229589 has spaces in the directory name, the URL to the local PDF should be escaped
- W8LS32717 has a funny character at the end of the path, see https://github.com/buda-base/archive-ops/issues/740

**Comment by JannTibetan** [comment]
> so, now two issues:
> 
> * what do we do with records that are not on the DLD but are found on BUDA? I think we want to have a clear distinction when the user accesses the images. In a first iteration this could be a popup: "This record is not on your local DLD drive, the images will be consulted on the BUDA website directly". Or perhaps through an icon?

I agree there should be a clear distinction between images on the drive and those available on the web. I think an icon might be confusing to some so let's have a text popup instead. 
Could we have an option for selecting between DLD-only search and full-archive search? This would necessitate search knowing the generation of the installed DLD but it sounds like we need to do that anyway (and if we're going to do that then we should take advantage of all the possibilities that follow from that.)

> * what to do with files that are on the DLD but are not supposed to be accessible on BUDA? Do we pretend they're not on the DLD? @JannTibetan

I'm torn😞... But as this is supposed to work in China we cannot have it access copyrighted texts. Please restrict them.

**Comment by berger-n** [comment]
> Could we have an option for selecting between DLD-only search and full-archive search? This would necessitate search knowing the generation of the installed DLD but it sounds like we need to do that anyway (and if we're going to do that then we should take advantage of all the possibilities that follow from that.)

I've added a facet (which is added automatically when starting search from a DLD) 
allowing user to focus on results available on the current DLD:

https://library.bdrc.io/scripts/offlinePDFaccess.html?url=https%3A%2F%2Flibrary.bdrc.io%2Fsearch%3Fq%3D%2522snying%2520thig%2520ya%2520bzhi%2522~1%26lg%3Dbo-x-ewts%26t%3DInstance%26f%3Dasset%2Cinc%2Ctmp%3ApossibleAccess%26f%3Dasset%2Cinc%2Ctmp%3AcatalogOnly%26%26f%3DinDLD%2Cinc%2Ctmp%3Aavailable

![image](https://user-images.githubusercontent.com/33383857/185622847-f8b1232e-1664-4f74-a0f9-5b31c592908f.png)
---

note that in the example above we are not on a real DLD, only 5 first volumes of W12827 are browsable, other volumes or scans won't be found although they are supposed to be on the DLD

also now the version of DLD is identified automatically (checking if files exist for some ids that are only available in one particular version of DLD)

**Comment by eroux** [comment]
This is really good, thanks a lot!

@JannTibetan I filtered out the copyrighted works from the DLD (and everything that is not open access, just in case)

I think we just need the popup for access online and then we're good... do you think we can already send a version to Karma? This is pretty advanced already

**Comment by eroux** [comment]
@karmagongde can you test the zip file https://github.com/buda-base/public-digital-library/files/9382433/BUDA-from-DLD.zip and see if it works for you?

**Comment by JannTibetan** [comment]
> This is really good, thanks a lot!
> 
> @JannTibetan I filtered out the copyrighted works from the DLD (and everything that is not open access, just in case)
> 

This really takes the DLD to the next level, thank you! It works really well, is easy to implement on a DLD, and most additional upgrades can be made purely on the backend.

**Comment by JannTibetan** [comment]
> @karmagongde can you test the zip file https://github.com/buda-base/public-digital-library/files/9382433/BUDA-from-DLD.zip and see if it works for you?

Dear Karma, thanks for testing this. I know you're on family vacation right now and probably a little jet lagged so we appreciate it. Since this is mainly for Tibetan users we want to make sure it is localized. Can you help us translate the translate: "Available on DLD"

@eroux Which other text needs to be translated?

**Comment by karmagongde** [comment]
Dear Jann, Elie, and Nicolas,

The new 00-index.html is not working for me. Please see the screenshot
below. The same title "thub pa'i dbang po'i skyes rabs gsal bar brjod pa
brgya" I searched on BUDA and got the search result.

"Available on DLD" can be translated as "སྟབས་བདེའི་དཔེ་མཛོད་ནང་ཐོབ།"

Best,
Karma

**Comment by eroux** [comment]
Hi @karmagongde thanks for trying! Unfortunately we can't see the screenshot, can you upload it on https://github.com/buda-base/public-digital-library/issues/734 ? thanks!

**Comment by karmagongde** [comment]
<img width="1635" alt="Screen Shot 2022-08-21 at 12 57 07 PM" src="https://user-images.githubusercontent.com/19337852/185787743-37255b9c-fb67-4405-82da-d8cf4b933a46.png">

**Comment by eroux** [comment]
Thanks, we'll look at it tomorrow. Just to make sure, can you try again in a private window?

**Comment by berger-n** [comment]
hi @karmagongde! is network ok? from your screenshot it seems it is not.

**Comment by JannTibetan** [comment]
>  "Available on DLD" can be translated as "སྟབས་བདེའི་དཔེ་མཛོད་ནང་ཐོབ།" Best, Karma

Thanks Karma

**Comment by karmagongde** [comment]
HI @berger-n, My network is ok. I am able to search texts on BUDA as well as able to upload and download texts through my network.

@eroux what is a private window and how to use the private window for "BUDA from DLD"?

**Comment by eroux** [comment]
if you click on the menu item on the right of the search / address bar in your web browser, you should see an option "open a private window", this will open a new window of your web browser that will not use what you already have in cache. Then you can just copy / paste the URL of the html file there

**Comment by berger-n** [comment]
> HI @berger-n, My network is ok. I am able to search texts on BUDA as well as able to upload and download texts through my network.

ok thanks! but in the screenshot above, you try to start search and nothing happens, is that it?
if so, can you please open inspector (⌥ Ctrl + ⌘ Cmd + i)
then go to network tab
and scroll all the way down to check if there's anything written in red?

![image](https://user-images.githubusercontent.com/33383857/185944218-03b72bff-c445-4d71-af8d-8d41dc29542d.png)

**Comment by berger-n** [comment]
hi again @karmagongde 
I've been able to reproduce the problem and fix it
you should only have to reload page (maybe in incognito mode first, as Elie suggested)

**Comment by berger-n** [comment]
there is now a popup asking user what to do when the pdf to open is not on DLD: https://library.bdrc.io/scripts/offlinePDFaccess.html?url=https%3A%2F%2Flibrary.bdrc.io%2Fsearch%3Fq%3D%2522snying%2520thig%2520ya%2520bzhi%2522~1%26lg%3Dbo-x-ewts%26t%3DInstance%26%26useDLD%3Dtrue%26pg%3D1%26f%3Dasset%2Cinc%2Ctmp%3ApossibleAccess%26f%3Dasset%2Cinc%2Ctmp%3AcatalogOnly%26f%3DinDLD%2Cexc%2Ctmp%3Aavailable

![image](https://user-images.githubusercontent.com/33383857/186205270-6beb6b45-4d19-4897-ab27-d04accf2ba41.png)
---

(buttons will use browser's locale but message needs to be translated)

**Comment by JannTibetan** [comment]
Thank you.

Let's modify the text to say:

This text is not available on your DLD.
Would you like to open it online instead?

**Comment by berger-n** [comment]
thanks @JannTibetan! 
done: https://library.bdrc.io/scripts/offlinePDFaccess.html?url=https%3A%2F%2Flibrary.bdrc.io%2Fsearch%3Fq%3D%2522snying%2520thig%2520ya%2520bzhi%2522~1%26lg%3Dbo-x-ewts%26t%3DInstance%26%26useDLD%3Dtrue%26pg%3D1%26f%3Dasset%2Cinc%2Ctmp%3ApossibleAccess%26f%3Dasset%2Cinc%2Ctmp%3AcatalogOnly%26f%3DinDLD%2Cexc%2Ctmp%3Aavailable

![image](https://user-images.githubusercontent.com/33383857/186461065-810101c3-40f9-4d08-add9-c551f2b1afc4.png)

**Comment by karmagongde** [comment]
“This text is not available on your DLD. Would you like to open it online instead?”
དེབ་འདི་ཁྱེད་ཀྱི་སྟབས་བདེའི་དཔེ་མཛོད་དུ་མི་འདུག། དེའི་ཚབ་དྲྭ་ཐོག་ཏུ་ཁ་འབྱེད་ཀྱི་ཡིན་ནམ།

**Comment by karmagongde** [comment]
BUDA-from-DLD,  
**00-index.html** could not identify DLD version. Please see the screenshot.
<img width="943" alt="Screen Shot 2022-09-02 at 9 28 54 PM" src="https://user-images.githubusercontent.com/19337852/188200259-c38c59cd-a5bf-447a-98d3-741f83c3cad5.png">

**Comment by berger-n** [comment]
> [BUDA-from-DLD.zip](https://github.com/buda-base/public-digital-library/files/9382433/BUDA-from-DLD.zip) contains a file and folder that should go at the top folder of the DLD

thanks @karmagongde! from your screenshot I think you need to move both the file `00-index.html` and the folder `99-DLDs` to the uppermost folder of the DLD drive (that is, move them near all the folders `01-ka-དཀར་ཆག` etc.)

so that when you open the DLD, `00-index.html` is the first name of the list (and `99-DLDs` the last at the bottom)
then if you open `00-index.html` the url is `/Volumes/DLD-2018/00-index.html` 
(it can't work if opened from a folder like `/BUDA-from-DLD/`)

**Comment by karmagongde** [comment]
Hi @berger-n , I tried to arrange 00-index.html and the folder 99-DLDs to the uppermost folder of the DLD drive by adding [a]. Still, it does not work. I will give you remotely control of my computer to you and please show me how it should be arranged.

**Comment by berger-n** [comment]
ok thanks @karmagongde! we'll need to check the Network tab of the inspector to see what's in red there

please open inspector (⌥ Ctrl + ⌘ Cmd + i) and click the Network tab at the top when it says "could not identify DLD" 
then instead of clicking "ok" just reload the page

after popup is displayed again, try to locate the line beginning with W00EGS1017030
then right-click on it (on mac, keep ⌥ Ctrl key down while clicking on it) and try to open it in a new tab:

![image](https://user-images.githubusercontent.com/33383857/188583810-9cd80a1e-a010-4b6b-a071-6bc764b69585.png)
---

now we need to know what url opens in browser, for me it is the following:

![image](https://user-images.githubusercontent.com/33383857/188587225-ddc15d97-5014-4d1a-9add-4281eb75d9cf.png)
---

we also need to know what files are actually displayed in Finder on DLD for this path (if it exists?):
 
`01-ka-དཀར་ཆག/01-སྤྱིར་བཏང་དཀར་ཆག/W00EGS1017030-ལྷ་བབས་མཆོད་རྟེན་ཆེན་མོའི་དཀར་ཆག་པདྨ་སྟོང་ལྡན།`

**Comment by karmagongde** [comment]
Hi @berger-n, is it a good time for you to check my network setting through remotely control?

**Comment by berger-n** [comment]
it is! let me send you a google meeting link
